File: iptvs/iptv/DeviceCallback.java
package com.yinhe.android.iptv;

import android.content.Context;
import android.util.Log;
import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import android.content.Intent;

@JNINamespace("content::iptv")
public class DeviceCallback {
    public interface OnVKSendListener {
        public boolean send();
    }

    private static final String TAG = "DeviceCallback";

    private static Context applicationContext = null;

    private static OnVKSendListener vkSendListener = null;

    public static void init(Context context) {
        if (context == null) {
            return;
        }
        if (applicationContext != null) {
            return;
        }

        applicationContext = context.getApplicationContext();
    }

    public static void setOnVKSendListener(OnVKSendListener listener) {
        vkSendListener = listener;
    }

    @CalledByNative
    private static boolean SendVK() {
        Log.d(TAG, "SendVK");
        if (vkSendListener != null) {
            vkSendListener.send();
        }
        return true;
    }
}

File: iptvs/iptv/device_callback.cc
#include "device_callback.h"
#include "jni/DeviceCallback_jni.h"
#include "algorithm"

namespace content {
namespace iptv {

std::vector<DeviceCallback::VKEventListener> DeviceCallback::VKEventListenerList;
std::vector<DeviceCallback::RefreshDisplayListener> DeviceCallback::RefreshDisplayListenerList;

void DeviceCallback::SendVK()
{
    for (std::vector<VKEventListener>::const_iterator iter = VKEventListenerList.begin();
        iter != VKEventListenerList.end(); ++iter) {
        (*iter)();
    }

    JNIEnv* env = base::android::AttachCurrentThread();
    Java_DeviceCallback_SendVK(env);
}

void DeviceCallback::addVKEventListener(VKEventListener listener)
{
    if (listener == NULL) {
        return;
    }

    if (std::find(VKEventListenerList.begin(), VKEventListenerList.end(), listener)
             == VKEventListenerList.end()) {
        VKEventListenerList.push_back(listener);
    }
}

void DeviceCallback::SetVideoWindow(int x, int y, int w, int h)
{
    for (std::vector<RefreshDisplayListener>::const_iterator iter =
            RefreshDisplayListenerList.begin();
            iter != RefreshDisplayListenerList.end(); ++iter) {
        (*iter)(x, y, w, h);
    }
}

void DeviceCallback::addRefreshVideoDisplayListener(RefreshDisplayListener listener)
{
    if (listener == NULL) {
        return;
    }

    if (std::find(RefreshDisplayListenerList.begin(), RefreshDisplayListenerList.end(), listener)
             == RefreshDisplayListenerList.end()) {
        RefreshDisplayListenerList.push_back(listener);
    }
}

}
}
